finality-aleph/src/substrate_network.rs

```diff
@@ -0,0 +1,413 @@
+                        }
+                    }
+                    NotificationsReceived { messages, .. } => {
+                        return Some(Messages(
+                            messages
+                                .into_iter()
+                                .filter_map(|(protocol, data)| {
+                                    match to_protocol(protocol.as_ref()) {
+                                        Ok(_) => Some(data),
+                                        // This might end with us returning an empty vec, but it's probably not
+                                        // worth it to handle this situation here.
+                                        Err(_) => None,
+                                    }
+                                })
+                                .collect(),
+                        ));
+                    }
+                    Dht(_) => continue,
+                },
+                None => return None,
+            }
+        }
+    }
+}
+
+impl<B: Block, H: ExHashT> Network for Arc<NetworkService<B, H>> {
+    type SenderError = SenderError;
+    type NetworkSender = SubstrateNetworkSender;
+    type PeerId = PeerId;
+    type Multiaddress = Multiaddress;
+    type EventStream = NetworkEventStream;
+
+    fn event_stream(&self) -> Self::EventStream {
+        Box::pin(self.as_ref().event_stream("aleph-network"))
+    }
+
+    fn sender(
+        &self,
+        peer_id: Self::PeerId,
+        protocol: Protocol,
+    ) -> Result<Self::NetworkSender, Self::SenderError> {
+        Ok(SubstrateNetworkSender {
+            // Currently method `notification_sender` does not distinguish whether we are not connected to the peer
+            // or there is no such protocol so we need to have this worthless `SenderError::CannotCreateSender` error here
+            notification_sender: self
+                .notification_sender(peer_id.into(), protocol_name(&protocol))
+                .map_err(|_| SenderError::CannotCreateSender(peer_id, protocol))?,
+            peer_id,
+        })
+    }
+
+    fn add_reserved(&self, addresses: HashSet<Self::Multiaddress>, protocol: Protocol) {
+        if let Err(e) = self.add_peers_to_reserved_set(
+            protocol_name(&protocol),
+            addresses
+                .into_iter()
+                .map(|address| address.into())
+                .collect(),
+        ) {
+            error!(target: "aleph-network", "add_reserved failed: {}", e);
+        }
+    }
+
+    fn remove_reserved(&self, peers: HashSet<Self::PeerId>, protocol: Protocol) {
+        let addresses = peers.into_iter().map(|peer_id| peer_id.0).collect();
+        self.remove_peers_from_reserved_set(protocol_name(&protocol), addresses);
+    }
+}
+
+impl<B: Block, H: ExHashT> NetworkIdentity for Arc<NetworkService<B, H>> {
+    type PeerId = PeerId;
+    type Multiaddress = Multiaddress;
+
+    fn identity(&self) -> (Vec<Self::Multiaddress>, Self::PeerId) {
```

Comment:
This function looks like LISP code :smile: Not saying it's necessarily a bad thing but... structs >> tuples